core/libraries/rest_api/RestIncomingQueryParamMetadata.php

```diff
@@ -0,0 +1,705 @@
+    }
+
+    /**
+     * Returns true if the value specified operator accepts arbitrary number of arguments, like "IN".
+     * @since $VID:$
+     * @param $operator
+     * @return bool
+     */
+    private function operatorIsNAry($operator)
+    {
+        $valueArray = $this->getQueryParamValue();
+        return array_key_exists(
+            $operator,
+            $this->getContext()->getModel()->valid_in_style_operators()
+        )
+            && isset($valueArray[1])
+            && is_array($valueArray[1])
+            && !isset($valueArray[2]);
+    }
+
+    /**
+     * Returns true if the operator accepts 3 arguments (eg "BETWEEN").
+     * @since $VID:$
+     * @param $operator
+     */
+    private function operatorIsTernary($operator)
+    {
+        $query_param_value = $this->getQueryParamValue();
+        return array_key_exists($operator, $this->getContext()->getModel()->valid_between_style_operators())
+            && isset($query_param_value[1])
+            && is_array($query_param_value[1])
+            && isset($query_param_value[1][0], $query_param_value[1][1])
+            && !isset($query_param_value[1][2])
+            && !isset($query_param_value[2]);
+    }
+
+    /**
+     * Returns true if the operator is a similar to LIKE, indicating the value may have wildcards we should leave alone.
+     * @since $VID:$
+     * @param $operator
+     * @return bool
+     */
+    private function operatorIsLike($operator)
+    {
+        $query_param_value = $this->getQueryParamValue();
+        return array_key_exists($operator, $this->getContext()->getModel()->valid_like_style_operators())
+            && isset($query_param_value[1])
+            && !isset($query_param_value[2]);
+    }
+
+    /**
+     * Returns true if the operator only takes one argument (eg it's like `IS NULL`).
+     * @since $VID:$
+     * @param $operator
+     * @return bool
+     */
+    private function operatorIsUnary($operator)
+    {
+        $query_param_value = $this->getQueryParamValue();
+        return array_key_exists($operator, $this->getContext()->getModel()->valid_null_style_operators())
+            && !isset($query_param_value[1]);
+    }
+
+    /**
+     * Returns true if the operator specified is a binary opeator (eg `=`, `!=`)
+     * @since $VID:$
+     * @param $operator
+     * @return bool
+     */
+    private function operatorisBinary($operator)
+    {
+        $query_param_value = $this->getQueryParamValue();
+        $model = $this->getContext()->getModel();
+        return isset($query_param_value[1])
+            && !isset($query_param_value[2])
+            && !array_key_exists(
+                $operator,
+                array_merge(
+                    $model->valid_in_style_operators(),
+                    $model->valid_null_style_operators(),
+                    $model->valid_like_style_operators(),
+                    $model->valid_between_style_operators()
+                )
+            );
+    }
```

Comment:
does NOT have to be done now, but these new methods will be super easy to write tests for now

Reply:
yeah thanks for the tip with leeway. Although for the moment I'm satisfied that they're covered by the unit test `ModelDataTranslator::testPrepareConditionsQueryParamsForModelsGood()` and `ModelDataTranslator::testPrepareConditionsQueryParamsForModelsBad()`.